src/webgldrawer.js

```diff
@@ -0,0 +1,1120 @@
+
+                    // bind buffers and set attributes before calling gl.drawArrays
+                    gl.bindBuffer(gl.ARRAY_BUFFER, this._secondPass.bufferTexturePosition);
+                    gl.vertexAttribPointer(this._secondPass.aTexturePosition, 2, gl.FLOAT, false, 0, 0);
+                    gl.bindBuffer(gl.ARRAY_BUFFER, this._secondPass.bufferOutputPosition);
+                    gl.vertexAttribPointer(this._firstPass.aOutputPosition, 2, gl.FLOAT, false, 0, 0);
+
+                    // Draw the quad (two triangles)
+                    gl.drawArrays(gl.TRIANGLES, 0, 6);
+
+                }
+
+                renderingBufferHasImageData = true;
+
+                if(useContext2dPipeline){
+                    // draw from the rendering canvas onto the output canvas, clipping/cropping if needed.
+                    this._applyContext2dPipeline(tiledImage, tilesToDraw, tiledImageIndex);
+                    renderingBufferHasImageData = false;
+                    // clear the buffer
+                    gl.bindFramebuffer(gl.FRAMEBUFFER, null);
+                    gl.clear(gl.COLOR_BUFFER_BIT); // clear the back buffer
+                }
+
+                // Fire tiled-image-drawn event.
+                // TODO: for discussion: do we need to ensure the image has been drawn to the output canvas already?
+                // Is it possible the image data may not be on the output canvas? Does it matter?
+                if( this.viewer ){
+                    /**
+                        * Raised when a tiled image is drawn to the canvas. Only valid
+                        * for webgl and canvas drawers.
+                        *
+                        * @event tiled-image-drawn
+                        * @memberof OpenSeadragon.Viewer
+                        * @type {object}
+                        * @property {OpenSeadragon.Viewer} eventSource - A reference to the Viewer which raised the event.
+                        * @property {OpenSeadragon.TiledImage} tiledImage - Which TiledImage is being drawn.
+                        * @property {Array} tiles - An array of Tile objects that were drawn.
+                        * @property {?Object} userData - Arbitrary subscriber-defined object.
+                        */
+                    this.viewer.raiseEvent( 'tiled-image-drawn', {
+                        tiledImage: tiledImage,
+                        tiles: tilesToDraw.map(info => info.tile),
+                    });
+                }
+
+            });
+
+            if(renderingBufferHasImageData){
+                this._outputContext.drawImage(this._renderingCanvas, 0, 0);
+            }
+
+        }
+
+        // Public API required by all Drawer implementations
+        /**
+        * Set the context2d imageSmoothingEnabled parameter
+        * @param {Boolean} enabled
+        */
+        setImageSmoothingEnabled(enabled){
+            //TODO: does imageSmoothingEnabled make sense in the WebGL drawer?
```

Comment:
I'm not sure. Do you know of a test image/scenario where this would be relevant? We could see if it makes a visually noticeable difference or not.

Reply:
My assumption is it wouldn't make a difference. The obvious test is to take a small image and scale it up a lot when drawing it. I feel like WebGL has its own controls over image interpolation, so it probably ignores `imageSmoothingEnabled`.

Reply:
So, the webgl drawer ultimately does copy data to a context2d canvas, and that canvas can have `imageSmoothingEnabled` set to `true`. However, all the scaling should have been done by this point, so I agree that this likely this doesn't matter. I'm happy to get rid of it but if we need to prove it is meaningless first we can create a test. I'm just not sure what it would involve to meaningfully test this.

Reply:
I don't think we need to prove it... It certainly makes sense that it wouldn't make a difference.

Reply:
Ok, I've made the function do nothing (as with the HTML drawer). Also, I removed the warning from the function in the HTML drawer - since it is always called from the `Viewer` constructor, that would lead to a warning each time it is called. The docs already mention this is only valid for canvas.